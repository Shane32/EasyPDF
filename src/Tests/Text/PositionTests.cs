using System.Drawing;
using Font = Shane32.EasyPDF.Font;

namespace Tests.Text;

public class PositionTests
{
    private readonly PDFWriter _writer = new PDFWriter();

    public PositionTests()
    {
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.NewPage(System.Drawing.Printing.PaperKind.Letter, false, 1f, 1f);
        _writer.PrepForTests();
    }

    [Theory]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1.2f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.18f, 0.03f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1f, 6f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0.083f, 1f, 0f, 0.875f)]
    [InlineData(1f, 0f, 1.2f, 0.1324f, 0.1366f, 0.043f, 0.18f, 0f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1.2f, 6f, 1.2f, 0.1324f, 0.1366f, 0.043f, 0.216f, 0.036f, 0.083f, 1f, 0f, 0.875f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1.2f, 0f, 1.05f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1f, 1f, 1.027f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1.2f, 1f, 1.233f)]
    public void TextSizes(float lineSpacing, float paragraphSpacing, float stretchY, float capHeight, float ascent, float descent, float height, float leading, float paragraphLeading, float stretchX, float characterSpacing, float width)
    {
        _writer.Font = new Font(StandardFonts.Times, 12f) {
            LineSpacing = lineSpacing,
            ParagraphSpacing = paragraphSpacing,
            StretchY = stretchY,
            CharacterSpacing = characterSpacing,
            StretchX = stretchX,
        };
        _writer.TextCapHeight().ShouldBe(capHeight, 0.001f);
        _writer.TextAscent().ShouldBe(ascent, 0.001f);
        _writer.TextDescent().ShouldBe(descent, 0.001f);
        _writer.TextHeight().ShouldBe(height, 0.001f);
        _writer.TextLeading().ShouldBe(leading, 0.001f);
        _writer.TextHeight().ShouldBe(_writer.TextAscent() + _writer.TextDescent() + _writer.TextLeading(), 0.001f);
        _writer.TextParagraphLeading().ShouldBe(paragraphLeading, 0.001f);
        _writer.TextWidth("Hello World!").ShouldBe(width, 0.001f);
    }
    
    [Fact]
    public void Standard()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedY()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedY_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f, LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedX()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { StretchX = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void CharacterSpacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { CharacterSpacing = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void MultiLine()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.5f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment + "\nLine 2");
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

}
