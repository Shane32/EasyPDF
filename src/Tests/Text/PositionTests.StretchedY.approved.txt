%PDF-1.4
%\xE2\xE3\xCF\xD3
2 0 obj
<</Length 917/Filter/FlateDecode>>stream
x\x9C\x8D\x97MO\xDC0\x10\x86\xEF\xF9\x159\xC2aS\x8F\x9D8qo\xA5j+\xA1^Js\xAB8T\xCBB\xA9X(t\xA5\xFE\xFD:3\x9E\x8F\xACXa\xF1\x91Y\xC73~\xF3`\xDEq\x9E\x9B\x8B\xB9\x09\xB1\x1D\x87\xD8\xCE7\xCD\xA7\xB9\xF9\xD6@\xEB\xF2\xD7\x06\xDA\xD1\xE7o\xD7n\xF7\x8Dk/\xF3\xCF\xEF\xC6u\xD0\xFEk~\\\xE7\xFB7\x8D\xC3yW_J\xF0r\xF7\xFA4\xEF]7\xE4h_"\xDF\xF5\xA9\xF50-\x97~\xF9\x0C\x11\xAF\xDB\xC6C\xE8\x06(\x83\xA0S\x01\xF3\x97\xFBK\xB4)\xA38w\xC3\x156\\\x02\xEB\xD20.W\xA6\x17\x11\xDB\xE6{\xC3\x05\xF7"\xED\x01GK\x91}\xC3\x8A\x96\xE1\xE7\xFC@\xF3v\xA1\xF4\xEE3\xB4C;\xDF2\x1F\x8F\xEB\x8E\x1DL\xED\xBC\xB7\x0C\xDCB\xF2\xC7\xD9\x87\x87\xFB\xBB\xC7\xFD\xEE\xF1\xF0\xFE|pg\xED\xD7\xDD\xEDa>\x07\xEF\xCE\x9E\xFE\x9C_\xCF\x97\xC4\x9A\x14\xE4\x12B\xC7\x19<\xDE\x97\xA7\xC3@\x00\x95a0\xD3\x81\xAA0#\x18p\x1EM_~c\x15\x0C\x84\x11\xD3\xFA4\x9D\xB5($#,\x87\x8C\x093\x09\x13J\xA9\xE1\xE4s\xB9T\xEAc\xFE\xB4{9\xC6\x94\xF7)\xAB\x09\x93\xDDE\x8E\xB7\x91[\xED#'\x98x:P\x15\xC6\x14\x82\xC1\x14\xA0T\xC1@0\xD10-\x1A\x14\x13VaLVX\x14L\x98Yv\x93\xAB\xC4\x94\xD3k\x19]\xFC\xFC\xBB{\xB8\x7F\xDC\x1DS\x1Az\x113DCi\x98\xCA\xF3a \x94\xCA0\x98\xE9@U\x98\xD2\x00\x86R\x9F\x98u\xB2\x94h\x98\x16\x05\xA5\x84U\x98\x92\x15\xD6\x0B%\xCC$J(\xA5\x86R^"\x85zPO\x87\xC3\xD3\xFE\x18S67V3\xF6\x06\xD3\x18\xCB\x03b \x98\xCA0\x98\xE9@U\x18SL\x06S\x1CK\x15\x0C\x04\x13\xE3\xA2e:kQLV\x98\x17L\x98I\x98P\xCA[\x98 \xF9\x0E\xB2\xECT\xEBO\xF4/w\xCA\xA1\x92\xDAe\xB2\x06\x9E\xD8\x7F\xD3\xCA\xC2\x93zx2&\x9E\xD4\xC5\xA7\xD1\xD0\x9A\x86R\x05\x03\xA1E\xC3\xB8h\x99\xCEZ\x94\x96\x15\xA6F\x8E\x99D+\xD589LS\x17\xB3\xECPkRD\xEBu\x9B\x02g\\\x13\xAC\x9F\x03\xB0\xA1S$\xBC\xF8\x06\xD8\x14(\xB5\xC4\xD4\xDD\xCA\xD5\x9D\xD8\xBA[\xFB\xBA3\xC6\xEE\xAC\xB3\xBB\x95\xB5[\x95\xCE\x98\xBBSw'Yo\xC3\xCB\x1B2`\x91zt\xA7\xDC\x0B\xBCz)x\xEB\xF2\x10\xD8\xE6)RxA\x8D^R\xA0\xD4\x12x\xDEz=x6{\x8A\x14\x9EW\xBB\xE7\x14\xD1e\xE0\xADT\xAA\xE5Sz\x81\x17jL\x7F\xD9y\x1E\x97\xAAt\xB4\x82\xEFUO\x83^-\x16zk\xFE\xD0\xB3\xFBS\xA4\xF0z\xF5\x7FI\x81RK\xE0\xF5\xB0:O$\xF9C\xAC\x9A@\xB9Q\x967m\x80j\x09\xBC\x95\xCA\xDE\x1C+\xB4\x15\x90\xAC7\xE1\x8D\xB9\xDEr|\x81Z\x97\xBB\xBA\xBF\xFBu\xFA\x18\x16\xD5z!\xDA\xA6\x00\x91\xBB\x02E\x8A/j_\x90\x14(\xB5\xF44f[\x03\x0C\xDC\x1B(R|\x83v\x07N\x11]\x06\xDFJ\xA5v\x08J/\xF8bU\x8F\xC8\x82\x871_\x86Z\xDBC|'\\or\xE6\xACh{\x04L\xDC$(Rx\x93\xB6\x09I\x81RK\xE0\x8D\xB6S\xC0\xC8\xAD\x82"\x857j\xB3\xE0\x14\xD1e\x0F\xB4V\xA56\x0CJ/\xF0\xA6\xAA\x96\x11]\x17\x12\x16\xA9&w\xD2\xF4\x92}\x01X\xBD\x018y\x05p\xEBw\x00g_\x02l\xC7H\xA6c\xA4U\xC7H\xD21\xD2\xBAc$\xD31\x92\xED\x18i\xD51\x8C\xCA%\x16v\xC9\xBC\x0F\xB8\xAA\x8E\x11C7\x06\\\xAB\xD2\xF5\x08\xDF\x91\xE9\xFD\x07$\xF6\x02

endstream
endobj
4 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]/Font<</F1 1 0 R>>>>/Contents 2 0 R/Parent 3 0 R>>
endobj
1 0 obj
<</Type/Font/Subtype/Type1/BaseFont/Helvetica/Encoding/WinAnsiEncoding>>
endobj
3 0 obj
<</Type/Pages/Count 1/Kids[4 0 R]/ITXT(3.3.2.0)>>
endobj
5 0 obj
<</Type/Catalog/Pages 3 0 R>>
endobj
6 0 obj
<<>>
endobj
xref
0 7
0000000000 65535 f 
0000001156 00000 n 
0000000015 00000 n 
0000001244 00000 n 
0000000999 00000 n 
0000001309 00000 n 
0000001354 00000 n 
trailer
<</Size 7/Root 5 0 R/Info 6 0 